server/models/user-page.js

```diff
@@ -0,0 +1,116 @@
+    var name = 'models.user-page.update';
+    _verbose(name + ' called', userId, url);
+    var lazyCreateParams = [uuid.v4(), userId, url, urlHash, title];
+    var lazyCreateUserPageQuery = 'WITH new_page AS (  ' +
+      '  INSERT INTO user_pages (id, user_id, url, raw_url, url_hash, title) ' +
+      '  SELECT $1, $2, $3, $3, $4, $5 ' +
+      '  WHERE NOT EXISTS (SELECT id FROM user_pages WHERE user_id = $2 AND url_hash = $4) ' +
+      '  RETURNING id ' +
+      ') SELECT id FROM new_page ' +
+      'UNION SELECT id FROM user_pages WHERE user_id = $2 AND url_hash = $4';
+    var addExtractedDataQuery = 'UPDATE user_pages SET ( ' +
+    'extracted_at, extracted_author_name, extracted_author_url, extracted_cache_age, ' +
+    'extracted_content, extracted_description, extracted_favicon_color, extracted_favicon_url, ' +
+    'extracted_image_caption, extracted_image_color, extracted_image_entropy, ' +
+    'extracted_image_height, extracted_image_url, extracted_image_width, extracted_language, ' +
+    'extracted_lead, extracted_media_duration, extracted_media_height, extracted_media_html, ' +
+    'extracted_media_type, extracted_media_width, extracted_provider_display, ' +
+    'extracted_provider_name, extracted_provider_url, extracted_published, extracted_safe, ' +
+    'extracted_title, extracted_type, extracted_url, updated_at) ' +
+    ' = ' +
+    '($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, ' +
+    '$19, $20, $21, $22, $23, $24, $25, $26, $27, $28, $29, $30) ' +
+    'WHERE user_id = $31 AND id = $32 ' +
+    'RETURNING *';
+    var addDataParams = [data.extractedAt, data.extractedAuthorName, data.extractedAuthorUrl, 
+      data.extractedCacheAge, data.extractedContent, data.extractedDescription,
+      data.extractedFaviconColor, data.extractedFaviconUrl, data.extractedImageCaption, 
+      data.extractedImageColor, data.extractedImageEntropy, data.extractedImageHeight, 
+      data.extractedImageUrl, data.extractedImageWidth, data.extractedLanguage, data.extractedLead,
+      data.extractedMediaDuration, data.extractedMediaHeight, data.extractedMediaHtml,
+      data.extractedMediaType, data.extractedMediaWidth,
+      data.extractedProviderDisplay, data.extractedProviderName, data.extractedProviderUrl,
+      data.extractedPublished, data.extractedSafe, data.extractedTitle, data.extractedType,
+      data.extractedUrl, new Date().toJSON(), userId];
+
+    _verbose('about to issue lazy user page creation query');
+    var userPageId;
+    postgres.query(lazyCreateUserPageQuery, lazyCreateParams)
+      .fail(userPage._onRejected.bind(userPage, name + ' failed early', cb))
+      .then(function(result) {
+        // we just got the page_id; push it onto the end of the params array
+        _verbose('the lazy create result is ' + JSON.stringify(result));
+        _verbose('the userPageId is ' + result.id);
+        userPageId = result.id;
+        addDataParams.push(result.id);
```

Comment:
I really dislike how clever this is. Maybe just define the params inside the callback.